xarray/core/computation.py

```diff
@@ -343,8 +344,10 @@ def apply_dataset_ufunc(func, *args, **kwargs):
     list_of_coords = build_output_coords(args, signature, exclude_dims)
 
     args = [getattr(arg, 'data_vars', arg) for arg in args]
+
     result_vars = apply_dict_of_variables_ufunc(
-        func, *args, signature=signature, join=join, fill_value=fill_value)
+        func, *args, signature=signature, join=data_vars_join,
```

Comment:
A couple of things:

1.) What do you think about removing the `join` kwarg in the `fillna` signature in both `dataarray.py` and `dataset.py`.  The writing of `ds0.fillna(ds1)` already implies a directionality, so adding a `join` option can lead to confusion.

2.) I've decided not to touch the default value of `dataset_fill_value=None`.  It leads to errors in many other places once you change that.   I'll look into raising some descriptive errors now that we have a `data_vars_join` option.  But if we want to make
```
out = add(ds0, ds2, 'outer', 'outer', dataset_fill_value=np.nan)
```
render
```
<xarray.Dataset>
Dimensions:  (x: 3)
Coordinates:
  * x        (x) int64 0 1 2
Data variables:
    a        (x) float64 1.0 2.0 nan
    b        (x) float64 nan 99.0 3.0
```
I think we'll have to do another PR.